src/os/PluggableDBStore.cc

```diff
@@ -0,0 +1,481 @@
+  int ret = load_symbols(g_conf->keyvaluestore_backend_library.c_str());
+  assert(ret ==0);
+  const char* path = g_conf->osd_data.data();
+  map<string, string> options;
+  if (PluggableDBStore::create_if_missing) {
+    options["create_if_missing"] = "true";
+  } else {
+    options["create_if_missing"] = "false";
+  }
+  int status = DBInit(path, getosdid(), &options);
+  assert(status);
+  init_done = true;
+  return 0;
+}
+
+int PluggableDBStore::do_open(ostream &out, bool create_if_missing)
+{
+  dout(10) << __func__ <<dendl;
+  if (!init_done) {
+    init(string());
+  }
+  PerfCountersBuilder plb(g_ceph_context, "db", l_PluggableDB_first, l_PluggableDB_last);
+  plb.add_u64_counter(l_PluggableDB_gets, "db_get");
+  plb.add_u64_counter(l_PluggableDB_txns, "db_transaction");
+  logger = plb.create_perf_counters();
+  cct->get_perfcounters_collection()->add(logger);
+  return 0;
+}
+
+PluggableDBStore::PluggableDBStore(CephContext *c) :
+  cct(c),
+  logger(NULL),init_done(false)
+{
+}
+
+PluggableDBStore::~PluggableDBStore()
+{
+  close();
+  dlclose(library);
+  delete logger;
+}
+
+void PluggableDBStore::close()
+{
+  //shutdown the DB backend
+  DBClose();
+  if (logger)
+    cct->get_perfcounters_collection()->remove(logger);
+}
+
+int PluggableDBStore::submit_transaction(KeyValueDB::Transaction t)
+{
+  dout(20) << __func__ <<dendl;
+  PluggableDBTransactionImpl * _t =
+    static_cast<PluggableDBTransactionImpl *>(t.get());
+
+  int ret = DBSubmitTransaction(_t->ops);
+  logger->inc(l_PluggableDB_txns);
+  return ret;
+}
+
+int PluggableDBStore::submit_transaction_sync(KeyValueDB::Transaction t)
+{
+  return submit_transaction(t);
+}
+
+void PluggableDBStore::PluggableDBTransactionImpl::set(
+  const string &prefix,
+  const string &k,
+  const bufferlist &to_set_bl)
+{
+  string key = combine_strings(prefix, k);
+  DBOp op;
+  op.type = DB_OP_WRITE;
+  string data;
+  to_set_bl.copy(0,to_set_bl.length(), data);
+  op.data = data;
+  ops[key] = op;
+}
+
+void PluggableDBStore::PluggableDBTransactionImpl::rmkey(const string &prefix,
+												         const string &k)
+{
+  string key = combine_strings(prefix, k);
+  DBOp op;
+  op.type = DB_OP_DELETE;
+  ops[key] = op;
+}
+
+void PluggableDBStore::PluggableDBTransactionImpl::rmkeys_by_prefix(const string &prefix)
+{
+  KeyValueDB::Iterator it = db->get_iterator(prefix);
+  for (it->seek_to_first();
+       it->valid();
+       it->next()) {
+    pair<string,string> key1 = it->raw_key();
+    string key =  key1.first + "\001" + key1.second;
+    DBOp op;
+    op.type = DB_OP_DELETE;
+    ops[key] = op;
+  }
+}
+
+int PluggableDBStore::get(
+    const string &prefix,
+    const std::set<string> &keys,
+    std::map<string, bufferlist> *out
+    )
+{
+  dout(30) << __func__ << dendl;
+  for (std::set<string>::const_iterator i = keys.begin();
+       i != keys.end();
+       ++i) {
+    string key = combine_strings(prefix, *i);
+	value_t *value = DBGetObject(key);
```

Comment:
I got it. "value_t" is a little misunderstand to me.

BTW since value_t is a cheap structure, maybe we can return the structure directly instead of pointer. free value later is a little tricky to me